anchor-service/src/merkle_tree.rs

```diff
@@ -0,0 +1,69 @@
+use anyhow::{anyhow, Result};
+use cid::Cid;
+
+use ceramic_core::DagCborIpfsBlock;
+use ceramic_event::anchor::{AnchorRequest, MerkleNode, MerkleNodes};
+
+pub struct MerkleTree {
+    pub root_cid: Cid,
+    pub nodes: MerkleNodes,
+    pub count: u64,
+}
+
+pub fn build_merkle_tree(anchor_requests: &[AnchorRequest]) -> Result<MerkleTree> {
+    if anchor_requests.is_empty() {
+        return Err(anyhow!("no requests to anchor"));
+    }
+    let mut peaks: [Option<Cid>; 64] = [None; 64];
+    let mut count: u64 = 0;
+    let mut nodes = MerkleNodes::default();
+    for anchor_request in anchor_requests {
+        count += 1;
+        // Ref: https://eprint.iacr.org/2021/038.pdf
+        // []
+        // [cid1]
+        // [none, (cid1, cid2)]
+        // [cid3, (cid1, cid2)]
+        // [none, none, ((cid1, cid2), (cid3, cid4))]
+        // [cid5, none, ((cid1, cid2), (cid3, cid4))]
+        // [none, (cid5, cid6), ((cid1, cid2), (cid3, cid4))]
+        // [cid7, (cid5, cid6), ((cid1, cid2), (cid3, cid4))]
+        // [none, none, none, (((cid1, cid2), (cid3, cid4)), ((cid5, cid6), (cid7, cid8))]
+        let mut new_node_cid: Cid = anchor_request.prev;
+        let mut place_value = 0;
+        while place_value < peaks.len() {
```

Comment:
I find this code confusing to understand.

Is there a reason we're using some fancy algorithm here instead of just the simple recursive balanced binary tree building algorithm taught in every college data structures class?  We know all the events up front, we don't need a streaming algorithm, and I think there's value in keeping this really simple and obviously correct for now.  If we want to move to a streaming approach to tree building we can do so in the future, but doing that would require changes to a lot more than just this function, so as long as we're building the whole tree in one go I'd strongly vote to keep the algorithm dead simple for now.

Reply:
Just spoke about this with @smrz2001 and he realized that because we need to replace older events with newer events on the same stream before we build the tree, that means the set of events that go into the tree isn't even strictly append-only.  Which means it's not even totally clear how we could *ever* do streaming tree building.  So it seems that using the MMR algorithm is not giving us any benefit now and is fairly likely to *never* give us any benefit.  So I fairly strongly think we should rip out this MMR approach and just put in the extremely simple balanced tree building algorithm instead.

Reply:
You mean something like 
```rs
fn build_merkle_tree_r(anchor_requests: &[AnchorRequest], nodes: &mut MerkleNodes) -> Result<Cid> {
    if anchor_requests.len() == 0 {
        return Err(anyhow!("Empty anchor_requests"));
    }
    if anchor_requests.len() == 1 {
        return Ok(anchor_requests[0].prev);
    }
    let top_power_of_2 = 1 << (63 - anchor_requests.len().leading_zeros());
    let split_point;
    if anchor_requests.len() == top_power_of_2 {
        split_point = top_power_of_2 / 2;
    } else {
        split_point = top_power_of_2;
    }
    let left = build_merkle_tree_r(&anchor_requests[..split_point], nodes)?;
    let right = build_merkle_tree_r(&anchor_requests[split_point..], nodes)?;
    let node = vec![Some(left),Some(right)];
    let node_cid = node.to_cid()?;
    nodes.insert(node_cid, node);
    Ok(node_cid)
}
```
That should build the same tree.